fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordStoreState.java

```diff
@@ -81,24 +85,53 @@
     protected final AtomicReference<RecordMetaDataProto.DataStoreInfo> storeHeader;
     @Nonnull
     protected final AtomicReference<Map<String, IndexState>> indexStateMap;
+    @Nonnull
+    protected final AtomicReference<Map<String, IndexMetaDataProto.IndexMetaData>> indexMetaDataMap;
 
     /**
-     * Creates a <code>RecordStoreState</code> with the given index states.
-     * Only indexes that are not in the default state ({@link IndexState#READABLE IndexState.READABLE})
-     * need to be included in the map.
+     * Creates a <code>RecordStoreState</code>.
+     *
+     * <p>
+     * While {@code indexStateMap} only includes indexes that are not in the default state ({@link IndexState#READABLE IndexState.READABLE}),
+     * {@code indexStateMap} includes all indexes.
+     * </p>
+     *
      * @param storeHeader header information for the given store
      * @param indexStateMap mapping from index name to index state
+     * @param indexMetaDataMap mapping from index name to index state meta-data
      */
     @API(API.Status.INTERNAL)
-    public RecordStoreState(@Nullable RecordMetaDataProto.DataStoreInfo storeHeader, @Nullable Map<String, IndexState> indexStateMap) {
-        final Map<String, IndexState> copy;
+    public RecordStoreState(@Nullable RecordMetaDataProto.DataStoreInfo storeHeader, @Nullable Map<String, IndexState> indexStateMap, @Nullable Map<String, IndexMetaDataProto.IndexMetaData> indexMetaDataMap) {
```

Comment:
Similar to my previous comment, perhaps this should validate that the metadata aligns with the state.  For example, that you don't have a state for an index that says it is readable, but the metadata says it is write-only.  Or, from my first comment, these could be combined.

Reply:
They are now combined.